### ISSUE TITLE
[cga] Support multiple streams in device_buffer and DevicePreallocatedAllocator

### ISSUE THREAD

**Issue by mimaric** [clara-parabricks/GenomeWorks#414]
`DevicePreallocatedAllocator::DeviceAllocate()` currently accept `cudaStream_t`. `DevicePreallocatedAllocator::DeviceFree()` waits on that stream before actually deallocating memory in order to prevent the memory from being deallocated before all work that uses it is done.
In cases when the same buffer is used by two or more streams it is not possible to wait on all of them and the user has to make sure that all other streams have finished.

The goal it to support associating one allocation with multiple streams.

This should also be supported by `device_buffer`/`buffer`.

Consider using variadic template to pass one or more streams, store them in a `vector` internally and loop over them in `DevicePreallocatedAllocator::DeviceFree()`

This issue is part of issue #318

**Comment by ahehn-nv** [comment]
Do we have a use case, where we need to associate multiple streams to a buffer, at this point?

I thought our current primary problem is that we encounter cases (during development) where a stream gets destroyed before the buffer associated with the stream gets deallocated - leading to a cudaStreamSynchronize on a freed/invalid stream handle. Is this in any way related or a entirely separate todo?

**Comment by ahehn-nv** [comment]
Just saw that we have a separate issue for the problem I described: #393 
So I guess this is unrelated. Could you elaborate a bit on where we need it?

**Comment by mimaric** [comment]
Yes, it is unrelated to that issue.

The reason we have associated streams are exception safety and RAII. Under normal circumstances user could sync before buffer gets destroyed, but RAII makes it much more convenient and exception safe.

The same holds for cases when the case when a buffer is associated with two or more streams.

One use case is overlapping fetching indices from host memory in one stream and processing and deallocating them in another (Issue #318). Again, if everything goes according to plan it would be enough to only sync on processing node (or not even there), but this way everything would be much safer.
